### PR TITLE
Add colocalization metrics

### PR DESCRIPTION
**Pull request by thanushipeiris**
## Description

This pull request implements several commonly used pixel-based colocalization metrics for fluorescent microscopy:

1. Pearson's Correlation Coefficient: pearson's correlation of pixel intensities across two matching channels
2. Manders' Colocalization Coefficient [paper](https://imagej.net/media/manders.pdf)
3. Manders' Overlap Coefficient [paper](https://imagej.net/media/manders.pdf)
4. Intersection Coefficients: fraction of a segmented channel that overlaps with another segmented channel
5. Pixel intensity sum: sum of pixel values within segmentation mask of a channel
6. Average pixel intensity: average pixel value within segmentation mask of a channel

A `roi` parameter is provided for each function for users to limit analysis to a particular region of interest.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/measure/__init__.py
Comment (mkcor) [inline skimage/measure/__init__.py:49]: A more descriptive function name would be better, e. g., `pearson_correlation_coefficient` or `pearson_corr_coeff`.

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:35]:
```suggestion
def pcc(image0, image1, roi=None):
```
Out of consistency throughout the API, see: https://scikit-image.org/docs/dev/api/skimage.metrics.html

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:27]:
```suggestion
            f"{arr} is of type {type(arr)}; should be ndarray of dtype bool.",
```
I can't see the need for the `name` argument... (?)

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:28]:
```suggestion
    elif bool_expected:
```

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:31]:
```suggestion
                f"{arr} is ndarray of dtype {arr.dtype} with non-binary ",
```

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:23]: I would just remove this check utility altogether. Perhaps just call something like

```Python
image = np.asarray(image)
```
for images and 
```
mask = np.asarray(mask, dtype=bool)
```
for masks

`asarray` will not make a copy if the input is already an array of the expected dtype

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:119]:
```suggestion
def mcc(image, mask, roi=None):
```

In cases with just one image and one mask, please just use `image` and `mask` as the parameter names.

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:273]: To avoid overhead, lets not generate `roi` in cases where it is not needed. 

```suggestion

image0 = np.asarray(image0)
image1 = np.asarray(image1)
if roi:
    roi = np.asarray(roi, dtype=bool)
    check_shape_equality_all(image0, image1, roi)
    image0 = image0[roi]
    image1 = image1[roi]
else:
    check_shape_equality_all(image0, image1)

denom = math.sqrt(np.sum(image0 * image0) * np.sum(image1 * image1))
return np.sum(image0 * image1) / denom
```

where an import of the math module will have to be done at the top.

Please also make similar stylistic changes to how ROIs are handled in other functions.

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:371]: I am not sure we should bother to provide dedicated functions for `av_pixel_intensity` or `pixel_intensity_sum`. That should be easy for users to get on their own. For example, this function is basically just `img[roi].mean()` with some shape checks.

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks for the contribution @thanushipeiris. Most of these seem reasonable to provide, but I think perhaps the sum and average cases are a bit too simple and we can let users just do those operations on their own.

I have left some initial comments, but this will need another round of review later. I did not yet look at the tests that were added.

File: skimage/measure/_colocalization.py
Comment (thanushipeiris) [reply skimage/measure/_colocalization.py]: I made this change for image arguments. For masks, I wanted to raise a warning if a non-binary mask was inputted (e.g. an intensity image layer) because if this is done accidently, the metric would be calculated without the user realising. I added the function `is_binary_ndarray()` in `skimage._shared.utils` to do that.

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:14]:
```suggestion
    r"""Calculate Pearson's Correlation Coefficient between pixel intensities in
```

Currently the equations are [not rendering properly](https://3469-2014929-gh.circle-artifacts.com/0/doc/build/html/api/skimage.measure.html#pearson-corr-coeff). I think adding `r` here should fix it. Alternatively, you can omit the `r` here, but will have to use double-backslashes like `\\frac` or `\\sum` in the math code below.

Comment (grlee77) [reply skimage/measure/_colocalization.py:14]: This should also resolve many of the flake8 messages like `Line 139:19: W605 invalid escape sequence '\s'` that the pep8 bot is complaining about.

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:143]:
```suggestion
        :math:`A_{i,coloc} = A_i` if :math:`Bmask_i > 0`
```

File: skimage/_shared/utils.py
Comment (grlee77) [inline skimage/_shared/utils.py:759]:
```suggestion
        if np.any((mask != 1) & (mask != 0)):
```
pep8 fix

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:96]:
```suggestion
        image0 = image0.reshape(-1)
        image1 = image1.reshape(-1)
```
reshape might be able to avoid a copy in some instances (see notes [here](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html#numpy.ravel))

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:88]: If we are going to call asarray on the images, we should probably also do it for `roi` before the call to `is_binary_ndarray`?

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:176]:
```suggestion
    s = np.sum(image)
    if s == 0:
        return 0
    return np.sum(image * mask) / s
```
(to avoid computing `np.sum(image)` twice)

Also tried using `np.sum(image[mask])` in place of `np.sum(image * mask)` but it is substantially slower

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:107]: I see that this metric relies on sums over the image. Do we need to check that `image.min() > 0` for this method? (i.e. was the method written for integer-valued images or can we use it for arbitrarily centered floating point images?).

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:256]: Similar question about if we need to check image0 & image1 for positive range?

Also, I only looked at the abstract, but this publication argues that use of MOC should be discouraged: https://doi.org/10.1002/cyto.a.24336. I suppose even if that is that case, having it may still be useful in reproducing results from previously published work.

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:279]:
```suggestion
        Fraction of `image0_mask` that overlaps with `image1_mask`.
```

File: skimage/measure/_colocalization.py
Comment (grlee77) [inline skimage/measure/_colocalization.py:268]:
```suggestion
        Image mask of channel A.
    image1_mask : (M, N) ndarray of dtype bool
        Image mask of channel B.
```

File: skimage/measure/tests/test_colocalization.py
Comment (grlee77) [inline skimage/measure/tests/test_colocalization.py:72]: I see that these tests are all for integer dtypes. If images must be non-negative, we should add a test for that. If negative values are allowed, then please add additional tests for that case, possibly with floating point inputs.

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks for making the initial updates. See some minor suggestions and a couple of questions about the valid image range for the Manders measures.

File: skimage/measure/_colocalization.py
Comment (jni) [inline skimage/measure/_colocalization.py:13]: I think `roi` should be renamed to `mask` to match other parts of the API?

File: skimage/measure/_colocalization.py
Comment (jni) [inline skimage/measure/_colocalization.py:61]:
```suggestion
    correlation. You may wish to plot the pixel intensities of each of the two
```

File: skimage/measure/_colocalization.py
Comment (jni) [inline skimage/measure/_colocalization.py:152]:
```suggestion
    References
    ----------
```

File: skimage/measure/_colocalization.py
Comment (jni) [reply skimage/measure/_colocalization.py:256]: I agree that it is still useful to have. Perhaps we can add that publication to the notes and references?

File: skimage/measure/_colocalization.py
Comment (jni) [inline skimage/measure/_colocalization.py:295]:
```suggestion
    nnz0 = np.count_nonzero(image0_mask)
    if nnz0 == 0:
        return 0
    nnz_joint = np.count_nonzero(np.logical_and(image0_mask, image1_mask))
    return nnz_joint / nnz0
```

File: skimage/measure/_colocalization.py
Comment (thanushipeiris) [reply skimage/measure/_colocalization.py:88]: added this to the `is_binary_ndarray()` function

File: skimage/measure/_colocalization.py
Comment (thanushipeiris) [reply skimage/measure/_colocalization.py:13]: Yeah but could this then be confusing for functions that use only masks like `intersection_coeff(image0_mask, image1_mask, roi=None)`? Does naming it `mask` provide enough of a distinction against those other `image0_mask` and `image1_mask` variables?

File: skimage/measure/_colocalization.py
Comment (thanushipeiris) [reply skimage/measure/_colocalization.py:256]: Yep I've actually already included this paper in the references and added a note (lines 216 - 243). Maybe I should turn that note into a warning?
```
    MOC does not take into account pixel intensities, just the fraction of
    pixels that have positive values for both channels[2]_ [3]_. Its usefulness
    has been criticized as it changes in response to differences in both
    co-occurence and correlation and so a particular MOC value could indicate
    a wide range of colocalization patterns [4]_ [5]_.

    References
    ----------
    .. [1] Manders, E.M.M., Verbeek, F.J. and Aten, J.A. (1993), Measurement of
           co-localization of objects in dual-colour confocal images. Journal
           of Microscopy, 169: 375-382.
           https://doi.org/10.1111/j.1365-2818.1993.tb03313.x
           https://imagej.net/media/manders.pdf
    .. [2] Dunn, K. W., Kamocka, M. M., & McDonald, J. H. (2011). A practical
           guide to evaluating colocalization in biological microscopy.
           American journal of physiology. Cell physiology, 300(4), C723–C742.
           https://doi.org/10.1152/ajpcell.00462.2010
    .. [3] Bolte, S. and Cordelières, F.P. (2006), A guided tour into
           subcellular colocalization analysis in light microscopy. Journal of
           Microscopy, 224: 213-232.
           https://doi.org/10.1111/j.1365-2818.2006.01
    .. [4] Adler J, Parmryd I. (2010), Quantifying colocalization by
           correlation: the Pearson correlation coefficient is
           superior to the Mander's overlap coefficient. Cytometry A.
           Aug;77(8):733-42.https://doi.org/10.1002/cyto.a.20896
    .. [5] Adler, J, Parmryd, I. Quantifying colocalization: The case for
           discarding the Manders overlap coefficient. Cytometry. 2021; 99:
           910– 920. https://doi.org/10.1002/cyto.a.24336
```

File: skimage/measure/_colocalization.py
Comment (thanushipeiris) [reply skimage/measure/_colocalization.py:107]: The [original paper](https://doi.org/10.1111/j.1365-2818.1993.tb03313.x) seems to implicitly assume pixel values are non-negative (e.g. in specifying the range of the metric as being from 0 to 1) so I'll add this.

File: skimage/measure/_colocalization.py
Comment (jni) [reply skimage/measure/_colocalization.py:13]: imho, yes, it does. :stuck_out_tongue_winking_eye:

**Comment by grlee77** [comment]
Hi @thanushipeiris, just checking in here. I added a small docstring formatting fix. Any chance you could add an example for the gallery?

**Comment by thanushipeiris** [comment]
Yep will do in the next couple of days, if I find a good set of images with colocalized stains

**Comment by alexdesiqueira** [comment]
@thanushipeiris thank you for all your work so far, it looks great! Would you like to add an example on the gallery? If you don't have time available, we could continue this PR from here. Thanks again!

**Comment by thanushipeiris** [comment]
@alexdesiqueira I've added an example now! Sorry for the delay, ended up just using blobs instead of actual image data.

**Comment by alexdesiqueira** [comment]
Everything looks green now :slightly_smiling_face: @grlee77 how is it now for you? Thanks! Thanks @thanushipeiris, too!

**Comment by alexdesiqueira** [comment]
Ping @grlee77 :slightly_smiling_face:

**Comment by grlee77** [comment]
Hi @thanushipeiris, I had forgotten about this one. The example looks pretty good to me, there was just one issue of a missing `plt.figure()` call that was causing the scatterplot to appear on top of the previous figure instead of in a separate one:

bad_figure

I pushed a fix for that and some other tiny pep8-related stuff. I think it is good now on my end.

Thank you for your patience.

**Review by grlee77 — APPROVED** [review]

**Comment by grlee77** [comment]
@jni or @mkcor, this seems ready to go after a 2nd review

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:40]:
```suggestion
    for perfect linear anti-correlation to +1 for perfect linear correlation.
```

File: skimage/measure/_colocalization.py
Comment (mkcor) [inline skimage/measure/_colocalization.py:66]:
```suggestion
    Providing the mask of only relevant sections of the image (e.g., cells, or
```
(pure convention)

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (grlee77) [reply doc/examples/applications/plot_colocalization_metrics.py:106]: The author surname is "Manders" with an s, so I think it is the former

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (grlee77) [inline doc/examples/applications/plot_colocalization_metrics.py:96]:
```suggestion
# Manders' Colocalization Coefficient (MCC)
```

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (grlee77) [inline doc/examples/applications/plot_colocalization_metrics.py:27]:
```suggestion
# The protein, "protein A", will be simulated as blobs and segmented.
```

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (grlee77) [inline doc/examples/applications/plot_colocalization_metrics.py:108]:
```suggestion
# In this image, while there are a lot of protein A spots within the nucleus
```

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (grlee77) [inline doc/examples/applications/plot_colocalization_metrics.py:129]:
```suggestion
# First, we will generate protein B and plot intensities of the two proteins in
```

**Review by grlee77 — COMMENTED** [review]
added a few more last minute grammatical fixes related to Marianne's comments. Given that these all are minor and seem non-controversial I will go ahead and accept the changes currently proposed here.

File: doc/examples/applications/plot_colocalization_metrics.py
Comment (mkcor) [reply doc/examples/applications/plot_colocalization_metrics.py:106]: Ok, thanks for checking. It's too bad there's a mistake in the title of one of the references! https://doi.org/10.1002/cyto.a.20896

File: skimage/_shared/utils.py
Comment (stefanv) [inline skimage/_shared/utils.py:659]: typo

**Review by stefanv — APPROVED** [review]
Approved, pending fixing of the typo.

File: skimage/_shared/utils.py
Comment (thanushipeiris) [reply skimage/_shared/utils.py:659]: It is actually spelled "explicitly" https://dictionary.cambridge.org/dictionary/english/explicitly

Comment (alexdesiqueira) [reply skimage/_shared/utils.py:659]: Hey @thanushipeiris,
that's what @stefanv was referring to, here :slightly_smiling_face: It now reads "explicitely," somehow:
>             "with bool data type. Please set order to 0 or explicitely "

Could you help us fixing it?
Thanks for your dedication to this PR! It looks like it's coming to an end. :slightly_smiling_face:

Comment (thanushipeiris) [reply skimage/_shared/utils.py]: ah my bad, I've tried to fix this now! 

Also, I tried merging some of the changes in the scikit image main branch which got a bit messy and has now added two new files: `tools/precompute/mc_meta/LookUpTable.h` and `dev.py` which I should not be making changes too...

**Comment by mkcor** [comment]
Is the addition of `.hypothesis/unicode_data/13.0.0/charmap.json.gz` intentional? :thinking:

**Review by lagru — COMMENTED** [review]
Just a small comment, this PR removes the executable permissions on `dev.py`. Probably useful to keep this. Did you do this intentionally? I think I saw a similar change elsewhere and wondering if git or some other tool is messing with us.

File: tools/precompute/mc_meta/LookUpTable.h
Comment (lagru) [inline tools/precompute/mc_meta/LookUpTable.h:2331]: I'm probably missing something, but right now it looks like this addition just duplicates the content of the file that's already there? The line length of the file is almost doubled exactly and I can't find a difference between the addition and what's already there...

Comment (thanushipeiris) [reply tools/precompute/mc_meta/LookUpTable.h]: Yep, I've deleted the duplicated second half and checked that it's now identical with the same file in the main branch of scikit-image.

**Comment by thanushipeiris** [comment]
> Is the addition of `.hypothesis/unicode_data/13.0.0/charmap.json.gz` intentional? 🤔

Nope, not sure where this came from and I have deleted it.


> Just a small comment, this PR removes the executable permissions on `dev.py`. Probably useful to keep this. Did you do this intentionally? I think I saw a similar change elsewhere and wondering if git or some other tool is messing with us.

As far as I can see, [this PR's dev.py](https://github.com/thanushipeiris/scikit-image/blob/colocalization-metrics/dev.py) is identical to the [scikit-image main branch](https://github.com/scikit-image/scikit-image/blob/main/dev.py). It's showing up as an "empty" file in the _Files changed_ section of this PR but in the actual repos it's identical I believe.

**Comment by thanushipeiris** [comment]
Also the tests are failing with `ModuleNotFoundError: No module named 'skimage.measure._colocalization'` even though the file `skimage.measure._colocalization.py` very clearly [exists](https://github.com/thanushipeiris/scikit-image/blob/colocalization-metrics/skimage/measure/_colocalization.py).

This has been happening for a long time on this PR and I'm not sure why or how to fix it...

File: tools/precompute/mc_meta/LookUpTable.h
Comment (lagru) [reply tools/precompute/mc_meta/LookUpTable.h]: No worries. Merge conflicts can get a bit messy. :sweat_smile:

**Comment by lagru** [comment]
> it's showing up as an "empty" file in the Files changed section of this PR.

I think GitHub's GUI is just confusing in this instance and the "Empty file." bit just means that there is no diff to show. If you look behind the [name of the file here](https://github.com/scikit-image/scikit-image/pull/6189/files#diff-e42998d51257e6f84aa51ffa5f4ed1544cb12f97a94978c44f3bc281b5324d79) you can see GitHub showing "100755 → 100644"; I guess that's why it shows up in this PR at all. Maybe a leftover from the merge conflict as well. 

Are you on Linux or similar? Then you can revert that bit with `chmod +x dev.py`. But I'm also happy to do it for you if you are on Windows. :slightly_smiling_face:

**Comment by lagru** [comment]
> Also the tests are failing with ModuleNotFoundError: No module named 'skimage.measure._colocalization' even though the file skimage.measure._colocalization.py very clearly [exists](https://github.com/thanushipeiris/scikit-image/blob/colocalization-metrics/skimage/measure/_colocalization.py).

We are switching to a new meson-based build system. The error should be resolved if you add the file `_colocalization.py`  to the [`meson.build`](https://github.com/scikit-image/scikit-image/blob/71da5143ca03d940291778dc37dd0f89856d1e85/skimage/measure/meson.build#L19) in the same directory.

**Review by jarrodmillman — APPROVED** [review]

**Comment by jarrodmillman** [comment]
@lagru This looks good to me. I am thinking about releasing 0.20.0rc0 tomorrow and wanted to check whether you thought we should merge this before or after 0.20. If before, feel free to merge.

**Comment by lagru** [comment]
I took the liberty to push a few last minor fixes and suggestions. I'll merge once the CI is green. Thank you @thanushipeiris and everyone else who was involved!

**Pull request merged**
